nilearn/glm/second_level/second_level.py

```diff
@@ -843,9 +843,25 @@ def non_parametric_inference(
     # Check design matrix and effect maps agree on number of rows
     _check_effect_maps(effect_maps, design_matrix)
 
+    # Obtain design matrix vars
+    var_names = design_matrix.columns.tolist()
+
     # Obtain tested_var
-    if contrast in design_matrix.columns.tolist():
+    if contrast in var_names:
         tested_var = np.asarray(design_matrix[contrast])
+        var_names.remove(contrast)
+
+    # Obtain confounding vars
+    if len(var_names) == 0:
+        # No other vars in design matrix
+        confounding_vars = None
+    else:
+        # Exclude the intercept (constant)
```

Comment:
I think that you need to explicitly need to remove the 'intercept' term ---if it's there, no ? Or test id the corresponding regressor is constant.

Reply:
Do you mean, "I don't think..."? 

If I do not remove it, and a user leaves the parameter `model_intercept=True`, the function `permuted_ols` will effectively add an intercept and also regress an intercept out of the data. This results in removing an extra degree of freedom. 

`permuted_ols` only checks if there is already an intercept in parameter `tested_vars` and not in `confounding_vars`. 

That's why I explicitly removed the intercept from the `confounding_vars`, if it's in the design matrix.

Reply:
I indeed think that we want to remove it. the point is that you are implicitly relying on the convention that if there is a unique regressor left, this is an intercept. I think that we should test
```if 'intercept' in var_names: ```

Reply:
Is the intercept always uniquely termed 'intercept'? In other words, users can't build design matrices where they give it an arbitrary name? (I haven't checked.)

Reply:
To my knowledge there is no constraint on intercept name. I think this is an okay way to test if the regressor is constant.

Reply:
On second thought could it be the case that `model_intercept` is set to False because a constant term is included and that this would remove it without the user knowing?

I'm wondering if it's better to add the check to the `permuted_ols` function here:
https://github.com/nilearn/nilearn/blob/9c29e3376243b643a72f78b8b98a56d12c2e99bb/nilearn/mass_univariate/permuted_least_squares.py#L696-L702

Reply:
I think adding the check to `permuted_ols` would make the most sense, since the same problem could come up there as well.